### ISSUE TITLE
GenomicsDBImport and GenotypeGVCFs produce a joint VCF missing the majority of calls

### ISSUE THREAD

**Issue by ksw9** [broadinstitute/gatk#5113]
Hi,
I am encountering an issue similar to what was described earlier (#3429)  in gatk-4.0.7.0, in which I find that the GenomicsDBImport combines gVCFs but leaves the majority of variants out of the final merged VCF. I called variants with DeepVariant I am not sure if the format is the problem as it was above. The header is below. Is there a fix for this or a temporary workaround?

Thank you!

##fileformat=VCFv4.2
##FILTER=<ID=PASS,Description="All filters passed">
##FILTER=<ID=RefCall,Description="Genotyping model thinks this site is reference.">
##FILTER=<ID=LowQual,Description="Confidence in this variant being real is below calling threshold.">
##INFO=<ID=END,Number=1,Type=Integer,Description="End position (for use with symbolic alleles)">
##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
##FORMAT=<ID=GQ,Number=1,Type=Integer,Description="Conditional genotype quality">
##FORMAT=<ID=DP,Number=1,Type=Integer,Description="Read depth">
##FORMAT=<ID=MIN_DP,Number=1,Type=Integer,Description="Minimum DP observed within the GVCF block.">
##FORMAT=<ID=AD,Number=R,Type=Integer,Description="Read depth for each allele">
##FORMAT=<ID=VAF,Number=A,Type=Float,Description="Variant allele fractions.">
##FORMAT=<ID=GL,Number=G,Type=Float,Description="Genotype likelihoods">
##FORMAT=<ID=PL,Number=G,Type=Integer,Description="Phred-scaled genotype likelihoods rounded to the closest integer">
##contig=<ID=NC_000962.3,length=4411532>

**Comment by droazen** [comment]
@ksw9 Can you check whether this is still an issue with the latest release (4.0.8.0)? That release introduced a new version of GenomicsDB. If it is still an issue in 4.0.8.0, can you provide a shareable test case to replicate the issue?

@kgururaj Could you comment on this one as well?

**Comment by kgururaj** [comment]
@ksw9  can you run the vcf validator tool suggested by @komalsrathi in #5045 [here](https://github.com/broadinstitute/gatk/issues/5045#issuecomment-407476343)? The issue is primarily caused by mismatch in the field description in the VCF header and the data lines.

@droazen can you comment on the [fix/check that I suggested here](https://github.com/broadinstitute/gatk/issues/5045#issuecomment-407501684)?

**Comment by kgururaj** [comment]
Do you agree with the actions for both the conditions listed in the comment?

**Comment by droazen** [comment]
I'm not sure about the padding aspect of the proposal...would client code need to be aware of the padding, or would it be purely internal to GenomicsDB?

**Comment by kgururaj** [comment]
Let's take the SB field as an example - the header says it's a fixed length field of 4 integers. What should happen if one of the data lines only contains 2 integers?

Similarly, let's say the AD field (length R) in a specific line contains fewer elements than the total number of alleles (alt+1). What should happen? 

This is irrespective of the internals of TileDB/GenomicsDB - should I report an error (exception) and exit? Print warning message, (pad if needed) and continue? Both those options are possible, it's a question of what is preferable.

**Comment by kgururaj** [comment]
The other scenario is what happens if the SB field in a line contains 6 integers?

**Comment by ldgauthier** [comment]
I'd rather be rigorous and error out.  If we pad the missing values there could still be other repercussions downstream.

**Comment by ksw9** [comment]
Dear all, thank you for the quick reply! Yes, I tried this on the latest release (4.0.8.0), and it is still an issue. vcf-validator validates the input VCFs, however, in the output VCF, I get errors associated with the field descriptions for the VAF annotation (which I believe may be specific to DeepVariant), a few of which are pasted below:

The header tag 'reference' not present. (Not required but highly recommended.)


column CDC1551_clean_mutated_9.fasta_1_1 at NC_000962.3:580772
\ .. FORMAT tag [VAF] expected different number of values (expected 1, found 2)
column CDC1551_clean_mutated_2.fasta_1_1 at NC_000962.3:580772 .. FORMAT tag [VAF] expected different number of values (expected 1, found 2)

I tried the above after changing the VAF field description label from Number=R to Number=A in my unmerged gVCF files. This did not solve the problem. I'm attaching two example VCF files which I am attempting to merge. My commands are below: 

```

# Create GenomicsDBImport
gatk GenomicsDBImport \
-R  ${REF_DIR}${ref} \
-V CDC1551_clean_mutated_1.fasta_1_1.bwa_deep.g.vcf.gz \
-V CDC1551_clean_mutated_2.fasta_1_1.bwa_deep.g.vcf.gz \
--reader-threads 8 \
--genomicsdb-workspace-path ${mapper}_${caller}_comb \
--intervals 'NC_000962.3' \
--java-options '-DGATK_STACKTRACE_ON_USER_EXCEPTION=true' \
--overwrite-existing-genomicsdb-workspace true 

## Joint Genotype VCF
gatk  GenotypeGVCFs \
-R ${REF_DIR}${ref} \
-V  gendb://${mapper}_${caller}_comb \
-O ${mapper}_${caller}_joint.vcf
```

[test_vcfs.zip](https://github.com/broadinstitute/gatk/files/2295977/test_vcfs.zip)


Thank you for your support on this problem!
Best,

**Comment by kgururaj** [comment]
I could retrieve the data correctly.

I "faked" the reference genome  - could you point me to reference fasta you are using?

**Comment by ksw9** [comment]
Thanks for checking this - I'm attaching the zipped reference fasta.
Best,
[H37Rv.fa.zip](https://github.com/broadinstitute/gatk/files/2307820/H37Rv.fa.zip)

**Comment by ksw9** [comment]
Yes, it runs for me as well, but, for me, output 37 variant sites. I have a true VCF for these files and the majority of known variants are missing. i.e. site 1977. For instance: 
zcat CDC1551_clean_mutated_1.fasta_1_1.bwa_deep.g.vcf.gz | grep 1977 

> NC_000962.3	1977	.	A	G,<NON_REF>	48.2	PASS	.	GT:GQ:DP:AD:VAF:PL	1/1:46:92:0,92,0:1,0:48,50,0,990,990,990

I'm worried that one of the annotations is unexpected by GATK and is inadvertently filtering out many true variants. Thanks for looking into this.

**Comment by kgururaj** [comment]
I see - can you run the tool SelectVariants instead of GenotypeGVCFs and check whether you get all the lines in the input gVCFs (I got them). If SelectVariants' output is correct, then GenomicsDB is functioning as expected.

**Comment by ksw9** [comment]
Hi,
Do you mean run SelectVariants on the output of gatk GenomicsDBImport?
I tried that. My GT calls are converted into nocalls for some reason. For example, a known variant site:

NC_000962.3     1977    .       A       G,<NON_REF>     .       .       .       GT:AD:GQ:PL:VAF ./.:0,92,0:46:48,50,0,990,990,990:1.00,0.00     ./.:0,90,0:46:47,49,0,990,990,990:1.00,0.00

I think this may be contributing to the problem.

**Comment by kgururaj** [comment]
Nope, that's expected behavior - SelectVariants uses a flag that disables GT field printing.

If you see all the expected lines in the output of SelectVariants (barring the GT field), then the data is stored and imported from GenomicsDB correctly.

**Comment by ksw9** [comment]
Okay, so that means that GenotypeGVCFs then is the issue, correct? Because the output of this call is a truncated set of variants which appear in my gVCFs.

**Comment by kgururaj** [comment]
GenotypeGVCFs is beyond my expertise - so I can't say anything

Returning to @droazen.

Summary: GenomicsDB stores and returns (using SelectVariants) all the variants as expected. @ksw9  has some questions about the output of GenotypeGVCFs.

**Comment by ldgauthier** [comment]
@ksw9 can you share the missing variants, i.e. those are are in the GenomicsDB output and not in the GenotypeGVCFs output?  I suspect that they're borderline in quality and aren't making the QUAL cutoff.  The fact that the PL[0] of the variant at NC_000962.3:1977 is only 48 means that the confidence that this size is not homozygous reference is pretty low.  This is just the likelihood, so when GenotypeGVCFs applies a prior of 0.001 (or -30 in Phred space) corresponding to a SNP mutation rate of one per thousand base pairs, this won't meet the emission threshold of 30 anymore.  Although it looks like that threshold got changed to 10 in GATK4, so that might not be the problem, but it would still be informative to look at the variants that get dropped.

**Comment by Jigdor** [comment]
@ksw9 has this issue been resolved?. I am using GenomicsDBImport and selectVariants (gatk/4.2.0.0) to combine gVCFs of 45 samples. At an individual sample gVCF, I see that none of the GTs are missing ("./.") but after I run GenomicsDBImport and then selectVariants, I see that all GTs in the combined gVCFs are set to "./.". Alternately, I tried using combinegVCF but I get the same "./." in the combinedgVCF.

**Comment by droazen** [comment]
@Jigdor The default for tools that read from GenomicsDB is to drop the genotypes. If you want to retain them, you need to run `SelectVariants`/`GenotypeGVCFs`/etc. with the `--call-genotypes` argument:

```
--call-genotypes <Boolean>    Output called genotypes in final VCF (otherwise no-call)  Default value: false. Possible
                              values: {true, false} 
```

**Comment by Jigdor** [comment]
@droazen Thanks for your help. It works great with that argument in.
